File: server/controllers/projects.py
from flask import Blueprint, request, jsonify
from pymongo import MongoClient
from bson.objectid import ObjectId
from middleware.auth import require_auth

bp = Blueprint("projects", __name__, url_prefix="/api/projects")
client: MongoClient = None

def init(mongo_client):
    global client
    client = mongo_client


def oid(v):
    return str(v) if isinstance(v, ObjectId) else v


def sanitize(doc):
    return {k: oid(v) if isinstance(v, ObjectId) else v for k, v in (doc or {}).items()}


@bp.get("")
@require_auth
def list_projects():
    db = client.get_database()
    projects = [sanitize(p) for p in db.projects.find({"ownerId": request.user["userId"]})]
    return jsonify(projects)


@bp.post("")
@require_auth
def create_project():
    db = client.get_database()
    data = request.json or {}
    doc = {
        "title": data.get("title"),
        "description": data.get("description"),
        "ownerId": request.user["userId"],
        "members": [],
        "managementMethod": data.get("managementMethod"),
    }
    res = db.projects.insert_one(doc)
    return jsonify({"_id": oid(res.inserted_id)}), 201


@bp.get("/<pid>")
@require_auth
def get_project(pid):
    db = client.get_database()
    p = db.projects.find_one({"_id": ObjectId(pid)})
    if not p:
        return jsonify({"error": "Not found"}), 404
    return jsonify(sanitize(p))


@bp.put("/<pid>")
@require_auth
def update_project(pid):
    db = client.get_database()
    data = request.json or {}
    db.projects.update_one({"_id": ObjectId(pid)}, {"$set": {
        "title": data.get("title"),
        "description": data.get("description"),
        "managementMethod": data.get("managementMethod"),
    }})
    p = db.projects.find_one({"_id": ObjectId(pid)})
    return jsonify(sanitize(p))


@bp.delete("/<pid>")
@require_auth
def delete_project(pid):
    db = client.get_database()
    db.projects.delete_one({"_id": ObjectId(pid)})
    db.tasks.delete_many({"projectId": pid})
    return jsonify({"ok": True})


@bp.post("/<pid>/members")
@require_auth
def add_member(pid):
    db = client.get_database()
    d = request.json or {}
    email = d.get("email")
    m = db.users.find_one({"email": email})
    if not m:
        return jsonify({"error": "User not found"}), 404
    db.projects.update_one({"_id": ObjectId(pid)}, {"$addToSet": {"members": str(m["_id"])}})
    return jsonify({"ok": True})


@bp.delete("/<pid>/members/<mid>")
@require_auth
def remove_member(pid, mid):
    db = client.get_database()
    db.projects.update_one({"_id": ObjectId(pid)}, {"$pull": {"members": mid}})
    return jsonify({"ok": True})

File: server/controllers/projects.js
import express from 'express';
import { Project } from '../models/Project.js';
import { Task } from '../models/Task.js';
import { User } from '../models/User.js';
import { requireAuth } from '../middleware/auth.js';

const router = express.Router();

// Get all projects for authenticated user
router.get('/', requireAuth, async (req, res) => {
  try {
    const projects = await Project.find({ ownerId: req.user.userId });
    res.json(projects);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Create project
router.post('/', requireAuth, async (req, res) => {
  try {
    const { title, description, managementMethod } = req.body;
    console.log('[projects] POST / - body:', JSON.stringify(req.body));
    const project = await Project.create({
      title,
      description,
      managementMethod: managementMethod,
      ownerId: req.user.userId,
      members: []
    });
    console.log('[projects] Created project managementMethod:', project.managementMethod);
    res.status(201).json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get project by ID
router.get('/:id', requireAuth, async (req, res) => {
  try {
    const project = await Project.findById(req.params.id);
    if (!project) {
      return res.status(404).json({ error: 'Project not found' });
    }
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Update project
router.put('/:id', requireAuth, async (req, res) => {
  try {
    const { title, description, managementMethod } = req.body;
    console.log(`[projects] PUT /${req.params.id} - body:`, JSON.stringify(req.body));
    console.log('[projects] PUT - has managementMethod key:', Object.prototype.hasOwnProperty.call(req.body, 'managementMethod'));
    console.log('[projects] PUT - managementMethod value/type:', managementMethod, typeof managementMethod);
    
    // Build update object with only defined fields
    const updateData = {};
    if (title !== undefined) updateData.title = title;
    if (description !== undefined) updateData.description = description;
    if (managementMethod !== undefined) updateData.managementMethod = managementMethod;
    
    const project = await Project.findByIdAndUpdate(
      req.params.id,
      updateData,
      { new: true, runValidators: true }
    );
    console.log('[projects] Updated project managementMethod:', project?.managementMethod);
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Delete project
router.delete('/:id', requireAuth, async (req, res) => {
  try {
    await Project.findByIdAndDelete(req.params.id);
    await Task.deleteMany({ projectId: req.params.id });
    res.json({ ok: true });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Add member to project
router.post('/:id/members', requireAuth, async (req, res) => {
  try {
    const { email } = req.body;
    const user = await User.findOne({ email });
    if (!user) {
      return res.status(404).json({ error: 'User not found' });
    }
    
    const project = await Project.findByIdAndUpdate(
      req.params.id,
      { $addToSet: { members: email } },
      { new: true }
    );
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Remove member from project
router.delete('/:id/members/:memberId', requireAuth, async (req, res) => {
  try {
    const project = await Project.findByIdAndUpdate(
      req.params.id,
      { $pull: { members: req.params.memberId } },
      { new: true }
    );
    res.json(project);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;
